XRTK.SDK/Packages/com.xrtk.sdk/Features/UX/Scripts/Controllers/Hands/BaseHandVisualizer.cs

```diff
@@ -0,0 +1,153 @@
+
+            MixedRealityHandTrackingProfile handTrackingProfile = MixedRealityToolkit.Instance.ActiveProfile.InputSystemProfile.HandTrackingProfile;
+            if (handTrackingProfile != null && !handTrackingProfile.EnableHandJointVisualization)
+            {
+                // clear existing joint GameObjects / meshes
+                foreach (var joint in joints)
+                {
+                    Destroy(joint.Value.gameObject);
+                }
+
+                joints.Clear();
+                return;
+            }
+
+            foreach (TrackedHandJoint handJoint in eventData.InputData.Keys)
+            {
+                Transform jointTransform;
+                if (joints.TryGetValue(handJoint, out jointTransform))
+                {
+                    jointTransform.position = eventData.InputData[handJoint].Position;
+                    jointTransform.rotation = eventData.InputData[handJoint].Rotation;
+                }
+                else
+                {
+                    GameObject prefab = MixedRealityToolkit.Instance.ActiveProfile.InputSystemProfile.HandTrackingProfile.JointPrefab;
+                    if (handJoint == TrackedHandJoint.Palm)
+                    {
+                        prefab = MixedRealityToolkit.Instance.ActiveProfile.InputSystemProfile.HandTrackingProfile.PalmJointPrefab;
+                    }
+                    else if (handJoint == TrackedHandJoint.IndexTip)
+                    {
+                        prefab = MixedRealityToolkit.Instance.ActiveProfile.InputSystemProfile.HandTrackingProfile.FingerTipPrefab;
+                    }
+
+                    GameObject jointObject;
+                    if (prefab != null)
+                    {
+                        jointObject = Instantiate(prefab);
+                    }
+                    else
+                    {
+                        jointObject = new GameObject();
+                    }
+
+                    jointObject.name = handJoint.ToString() + " Proxy Transform";
+                    jointObject.transform.position = eventData.InputData[handJoint].Position;
+                    jointObject.transform.rotation = eventData.InputData[handJoint].Rotation;
+                    jointObject.transform.parent = transform;
+
+                    joints.Add(handJoint, jointObject.transform);
+                }
+            }
+        }
+
+        public void OnMeshUpdated(InputEventData<HandMeshUpdatedEventData> eventData)
+        {
+            if (eventData.Handedness != Controller?.ControllerHandedness)
+            {
+                return;
+            }
+
+            if (handMeshFilter == null &&
+                MixedRealityToolkit.Instance.ActiveProfile.InputSystemProfile.HandTrackingProfile?.HandMeshPrefab != null)
+            {
+                handMeshFilter = Instantiate(MixedRealityToolkit.Instance.ActiveProfile.InputSystemProfile.HandTrackingProfile.HandMeshPrefab).GetComponent<MeshFilter>();
```

Comment:
A visualiser should not be generating/instantiating meshes, that is the role of a controller/provider, to deliver content to visualise.  The visualiser should only be managing the display and manipulation of the data it is given.